libgrust/libproc_macro/ident.cc

```diff
@@ -0,0 +1,71 @@
+// Copyright (C) 2023 Free Software Foundation, Inc.
+//
+// This file is part of the GNU Proc Macro Library.  This library is free
+// software; you can redistribute it and/or modify it under the
+// terms of the GNU General Public License as published by the
+// Free Software Foundation; either version 3, or (at your option)
+// any later version.
+
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// Under Section 7 of GPL version 3, you are granted additional
+// permissions described in the GCC Runtime Library Exception, version
+// 3.1, as published by the Free Software Foundation.
+
+// You should have received a copy of the GNU General Public License and
+// a copy of the GCC Runtime Library Exception along with this program;
+// see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
+// <http://www.gnu.org/licenses/>.
+#include "ident.h"
+
+#include <cstring>
+
+extern "C" {
+Ident
+Ident__new (unsigned char *str, std::uint64_t len)
+{
+  unsigned char *val = new unsigned char[len];
+  std::memcpy (val, str, len);
+  return {false, val, len};
+}
+
+Ident
+Ident__new_raw (unsigned char *str, std::uint64_t len)
+{
+  unsigned char *val = new unsigned char[len];
+  std::memcpy (val, str, len);
+  return {true, val, len};
+}
+
+void
+Ident__drop (Ident *ident)
+{
+  delete ident->val;
```

Comment:
It is indeed undefined behavior to delete an array with `delete` instead of `delete[]`, no matter how strong is your CPP-fu, it seems like mine is weaker.